src/commands/test/runner.py

```diff
@@ -150,22 +166,45 @@ async def invoke_test_function(self, function_name: str):
         )
 
         func = getattr(self.test_contract, function_name)
+        is_failure_expected = (
+            function_name.startswith("test_fail_")
+            if self._is_test_fail_enabled
+            else False
+        )
         # TODO: Improve stacktrace
         try:
-            call_result = await func().invoke()
-            if self._is_test_error_expected:
-                raise MissingExceptReportedException(
-                    "Expected a transaction to be reverted"
+            try:
+                call_result = await func().invoke()
+                if self._expected_error is not None:
+                    raise MissingExceptReportedException(
+                        "Expected a transaction to be reverted"
+                    )
+                if is_failure_expected:
+                    raise TestNotFailedReportedException()
+                return call_result
+
+            except StarkException as ex:
+                is_ex_unexpected = (
+                    self._expected_error is None
+                    or re.compile(self._expected_error.name).search(ex.code.name)
+                    is None
+                    or re.compile(self._expected_error.message).search(ex.message or "")
+                    is None
                 )
-            return call_result
 
-        except StarkException as ex:
-            if not self._is_test_error_expected:
-                raise StarkExceptionReportedException(ex) from ex
+                if is_ex_unexpected:
+                    raise StarkExceptionReportedException(ex) from ex
 
+                if is_failure_expected:
+                    raise TestNotFailedReportedException() from ex
+        except TestNotFailedReportedException as ex:
+            raise ex
```

Comment:
Why do we catch it?

Reply:
To avoid `TestNotFailedReportedException` being caught by the code below. Otherwise, `TestNotFailedReportedException` would never be thrown from that function.